runtime/near-vm-logic/src/logic.rs

```diff
@@ -943,14 +943,200 @@ impl<'a> VMLogic<'a> {
+        rounds: u32,
+        state_len: u64,
+        state_ptr: u64,
+        message_len: u64,
+        message_ptr: u64,
+        t0: u64,
+        t1: u64,
+        f0: u64,
+        f1: u64,
+        register_id: u64,
+    ) -> Result<()> {
+        use blake2::VarBlake2b;
+        use std::convert::TryFrom;
+
+        if state_len != 8 {
+            return Err(HostError::Blake2InvalidStateLength { length: state_len }.into());
+        }
+
+        // Change to per block for gas.
+        let message_blocks = std::cmp::max(message_len / 128, 1);
+
+        self.gas_counter.pay_base(blake2b_base)?;
+        self.gas_counter.pay_per(blake2b_block, message_blocks)?;
+        self.gas_counter.pay_per(blake2b_round, (rounds as u64) * message_blocks)?;
+
+        let state = <[u64; 8]>::try_from(self.memory_get_vec_u64(state_ptr, 8)?)
+            .expect("vec bytes conversion");
+        let m = self.memory_get_vec(message_ptr, message_len)?;
+
+        let mut hasher = VarBlake2b::with_state(rounds, state, [t0, t1]);
+        hasher.update_inner(&m);
+        hasher.compress(f0, f1);
+        let res = hasher.output();
+
+        self.internal_write_register(register_id, res.as_slice().to_vec())
+    }
+
+    /// Hashes the given message using BLAKE2s and returns it into `register_id`.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// If the message is too long and overflows, an error is returned with
+    /// `Blake2HashDataOverflow`.
+    ///
+    /// # Cost
+    ///
+    /// Where `message_blocks` is `max(message_len / 128, 1)`.
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes +
+    /// message_blocks * blake2s_block + rounds * message_blocks`
+    #[cfg(feature = "protocol_feature_evm")]
+    pub fn blake2s(
+        &mut self,
+        rounds: u32,
+        state_len: u64,
+        state_ptr: u64,
+        message_len: u64,
+        message_ptr: u64,
+        t: u64,
+        f0: u32,
+        f1: u32,
+        register_id: u64,
+    ) -> Result<()> {
+        use blake2::VarBlake2s;
+        use std::convert::TryFrom;
+
+        if state_len != 8 {
+            return Err(HostError::Blake2InvalidStateLength { length: state_len }.into());
+        }
+
+        // Change to per block for gas.
+        let message_blocks = std::cmp::max(message_len / 64, 1);
+
+        self.gas_counter.pay_base(blake2b_base)?;
+        self.gas_counter.pay_per(blake2b_block, message_blocks)?;
+        self.gas_counter.pay_per(blake2b_round, (rounds as u64) * message_blocks)?;
+
+        let state = {
+            let mut buf = [0u32; 8];
+            let values = self.memory_get_vec_u64(state_ptr, 8)?;
+            for (x, y) in buf.iter_mut().zip(values.into_iter()) {
+                *x = <u32>::try_from(y).map_err(|_| HostError::Blake2HashDataOverflow)?;
```

Comment:
Why `HostError::Blake2HashDataOverflow` needs to exist? Seems like we can ask the caller to pass  a bunch of `u32`, rather than asking them to pass a bunch of `u64` and then checking if those `u64`s are actually `u32`s?